### PR TITLE
fix(streams): skip missing gauntlet probes in the recovery EMA

### PR DESCRIPTION
**Pull request by ss251**
## Outcome

Fix.

`steps_to_criterion` / `gauntlet_scorecard` (documented public gauntlet API in `alberta_framework.streams.gauntlet`) silently reported the segment-length "never recovered" cap after a later finite recovery, because one mid-segment NaN poisoned the rest of the causal EMA.

Supported path: `from alberta_framework.streams.gauntlet import steps_to_criterion, gauntlet_scorecard`. NaN in squared-error traces is first-class: the same scorecard documents `nan_steps` (P6) on the same array.

On live origin/main `c9aba7b54dedd647f8bd5f5c7bf6780b1413b676`, this 421-step segment (`decay = 0.5**(1/50)`):

```
[1.0] * 20 + [1.0] + [0.0] * 400   -> steps_to_criterion(..., 0.05) = 237
[1.0] * 20 + [nan] + [0.0] * 400   -> steps_to_criterion(..., 0.05) = 421  (cap = len)
```

The poison trace has 20 finite EMA values and last EMA `nan`. After the one-boundary fix, the same poison call returns `237` (last EMA ~0.0039). `gauntlet_scorecard` on the documented nine-segment program (`segment_length=421`, poison in segment 2) reports `recovery_steps_c=303` and `nan_steps=1` instead of the never-recovered cap `421`.

All-finite traces stay bit-identical (`test_ema_smooth_preserves_leading_batch_axes`). Load-bearing: reverting only `ema_smooth` to origin makes both new tests fail `421 < 421`.

This is not a Kayvan skip-zero / exact-dict series, not a 10000-cap / walk-bound sibling of #2697, not a JEPA late-return glance-slice of #2696, not metrics.py FWT / tracking-error (#2694/#2695), and not leftover-tax (CanonicalUPGD / feature_discovery / clocks / forager / IA / FTL / upgd / horde / dreaming / delight / #1874 / #2157-#2177). No open ASI PR titles the same hole.

## Measurement gate

- Lane: documented gauntlet API `steps_to_criterion` / `gauntlet_scorecard`
- Metric: steps-to-criterion / `recovery_steps_c` after a missing probe
- Origin proof at `c9aba7b54dedd647f8bd5f5c7bf6780b1413b676`: poison `421` (never recovered)
- Overlay at this head: poison `237` (recovers with the all-finite twin)
- Focused pytest `tests/test_gauntlet_recovery_missing_probes.py` + existing EMA / scorecard-window tests: 7 passed
- `ruff check` on the two changed files: clean
- One production file: `alberta_framework/streams/gauntlet.py`

- [x] logs: N/A - focused unit tests on the documented gauntlet API; no benchmark shard was run

- [x] domain-artifact: N/A - no `outputs/**` artifact; the measurement is steps-to-criterion `421` vs `237` on the traces above

- [x] llm-trajectory: N/A - official `run-receipt.mjs start` failed before trace: `--client must be a concrete identifier`; this checkout origin is elizaOS/asi not SlopDotCash/asi; shared primary remotes were not mutated

- AI assistance: yes

- Model(s) used: xai/grok-4.6

- Agent tooling: Grok Bot.app

- Skill path: SlopDotCash/slopdotcash@92d692518c3d832ac9b203076ef691a54e61a9fa:skills/contribute-to-asi

- Provenance status: self-reported

AI provider/model: xAI / grok-4.6
Client / agent tooling: Grok Bot.app
Contribution skill revision: SlopDotCash/slopdotcash@92d692518c3d832ac9b203076ef691a54e61a9fa:skills/contribute-to-asi
Attribution status: self-reported
— [grok-fleet-ss251]

### CODE REVIEW

**Review by saltwire-dev — CHANGES_REQUESTED** [review]
## Request changes at exact head `a628f22c261cac1af386c1bce434a5f18b8b6380`

Reviewed at exact head `a628f22c261cac1af386c1bce434a5f18b8b6380` against production `alberta_framework/streams/gauntlet.py` (not just the new test). Conversation sidebar is No reviews; the github-actions “produced no report” note is a comment, not a review. CPython 3.13.5, jax 0.11.1, numpy 2.2.4.

Skipping a mid-segment NaN so the carry stays at the last finite EMA is the right *direction* for the PR body’s 20-ones / one-NaN / 400-zeros row — on this CPU origin `steps_to_criterion` reports **421** (never-recovered cap) and this head reports **237**, matching the all-finite twin, with last EMA `≈0.0039`. The hole is that the gate is **`jnp.isfinite`**, so **±Inf is treated as a missing probe**, and a **non-finite prefix is the scan init carry**. `finite_c` is then False, and the first finite 0 **restarts the EMA at 0.0**. A learner that diverged to `+inf` for 20 steps then went quiet is scored recovered at step **20**. The same shape with a finite `1e20` prefix — still a legal float32 squared error, still catastrophic — never recovers (cap **420**). The new tests lock only the unique mid-NaN-among-ones fixture; the scorecard test is `rec_c < length`.

**Blocking: `isfinite` + non-finite init carry restarts the EMA after leading ±Inf; finite overflow does not**

`jax.lax.scan(..., values[..., 0], time_major)` still seeds the carry with the **raw first observation**. When that value is `+inf` / `-inf` / `NaN`, `finite_c` is False for every later step until a finite `v` arrives; the `jnp.where(finite_c, updated, v)` branch then sets `new_carry = v`. Squared error of `+inf` is a diverged learner, not an unevaluated probe. `nan_steps` (P6) uses the same `~isfinite` predicate, so Inf and NaN are both already flagged — and the recovery clock still treats Inf as “not evaluated.”

On this CPU (threshold `0.05` unless noted; traces float32, length 421):

| input | new `steps_to_criterion` | origin |
|---|---|---|
| PR live test: 20×`1` + `nan` + 400×`0` | **237** | **421** cap |
| 20×`1` + `+inf` + 400×`0` | **237** (Inf erased like NaN) | 421 |
| 20×`1` + `1e20` + 400×`0` | **421** never | 421 |
| **20×`+inf` + 400×`0`** | **20** (restart at 0.0) | 420 |
| 20×`1e20` + 400×`0` | **420** cap | 420 |
| scorecard thr `0.02` on PR poison | **303** | 421 |

`20 × +inf` then zeros is a supported squared-error trace. After this PR the recovery clock says the learner hit criterion at the first quiet step. The finite-`1e20` neighbor still says never. Mid-segment Inf is the same erase-vs-keep split. A missing-probe fix should not make overflow look like a logging hole.

`gauntlet_scorecard` on the PR’s own poison-in-segment-2 array still reports **`early_mse_c_first = NaN`** and **`savings_c = NaN`**, because index 20 sits inside the default `early_window_mse(..., window=200)` and those helpers still `jnp.mean` the raw window. P3 recovery is now a number; P4/P5 memory metrics on the same missing probe are still undefined.

**Tests lock the unique mid-NaN-among-ones fixture, not Inf, not a leading hole, not the scorecard threshold**

Only `jnp.nan`, never `+inf` / `-inf` / finite overflow. Only a **mid** hole among **ones**. Leading 20 NaN/Inf then zeros is untested; on this CPU it reports **20**, which still satisfies `rec_c < length`. Scorecard assertion is **`rec_c < 421`**, not `== 303`. Threshold `0.05` in the unit test is **not** the scorecard’s `2 * noise_floor = 0.02`.

**What I would approve**

Keep skipping **NaN**. Then: (1) gate on `jnp.isnan`, not `jnp.isfinite` — `+inf` / `-inf` stay in the carry (or force the never-recovered cap); (2) do not restart from a non-finite init carry into the first quiet 0; (3) pin 20×`+inf` + 400×`0` → **cap / never**, not 20; pin scorecard poison-in-seg-2 `recovery_steps_c == 303` not just `< 421`.

Until `steps_to_criterion` on `[+inf]*20 + [0]*400` either rejects or reports never-recovered rather than **20**, this does not lock “skip missing probes” without also scoring overflow as recovered.

**Comment by lalalune** [comment]
Closing after review: this treats every non-finite error, including infinity from divergent learning, as an unevaluated probe and allows a later recovery verdict. The gauntlet error trace does not carry an explicit missing-probe mask. Conflating divergence with missing measurements weakens honest failure accounting; a missing-probe feature needs an explicit protocol and separate treatment of infinity.
